frappe/doctypes.py

```diff
@@ -0,0 +1,53 @@
+from typing import Dict, Type
+from frappe.model.document import Document
+from frappe.model.base_document import get_controller
+
+import frappe
+
+FRAPPE_KEY = bytes("frappe", "utf8")
+
+
+def modulize(dt: str) -> str:
+	"""Convert string into importable Python namespace
+	"""
+	return dt.replace(" ", "").replace("-", "_")
+
+
+def generate_doctype_map() -> Dict[str, Type[Document]]:
+	"""Generate and return DocType -> Module Controller mapping
+	"""
+	doctype_mapping = {}
+	modules = frappe.get_all("Module Def", {"app_name": "frappe"}, pluck="name")
+	doctypes = frappe.get_all(
+		"DocType", {"module": ("in", modules), "custom": False}, pluck="name"
+	)
+
+	for dt in doctypes:
+		dt_ct = modulize(dt)
+		try:
+			doctype_mapping[dt_ct] = get_controller(dt)
+		except ImportError:
+			continue
+	return doctype_mapping
+
+
+def generate():
+	"""Set bench_controllers cache in Redis
+	"""
+	if frappe.cache().hgetall("bench_controllers").get(FRAPPE_KEY):
+		return
+
+	frappe.cache().hget(
+		"bench_controllers", "frappe", generator=generate_doctype_map, shared=True
+	)
+
+
+try:
+	BENCH_CONTROLLERS = frappe.cache().hgetall("bench_controllers").get(FRAPPE_KEY, {}) or {}
+except Exception:
+	BENCH_CONTROLLERS = generate_doctype_map()
+
+
+# loads controllers from Redis and save it under the frappe.doctypes namespace
+for dt, ct in BENCH_CONTROLLERS.items():
+	globals()[modulize(dt)] = ct
```

Comment:
Controller class already available in globals(), seems it can be directly accessed like below
(no need to import from frappe.doctypes)
e.g
 `User('User','Administrator')`

Also please consider adapt the Document __init__ method to set self.doctype= type(self).__name__ if args length is 1 and type is string
```
class Document(BaseDocument):
	"""All controllers inherit from `Document`."""
	def __init__(self, *args, **kwargs):
		"""Constructor.

		:param arg1: DocType name as string or document **dict**
		:param arg2: Document name, if `arg1` is DocType name.

		If DocType name and document name are passed, the object will load
		all values (including child documents) from the database.
		"""
		self.doctype = self.name = None
		self._default_new_docs = {}
		self.flags = frappe._dict()

		if args and args[0] and isinstance(args[0], string_types):
			# first arugment is doctype
			if len(args)==1:
				# single
				self.doctype = self.name = args[0]
```
then finally pass only docname to Controller class to get the doc. 
`User('Administrator')`

Reply:
> Also please consider adapt the Document init method to set self.doctype= type(self).name if args length is 1 and type is string

check https://github.com/frappe/frappe/pull/13223